File: android/app/src/main/kotlin/com/baseapp/android/view/games/astro/AstroEngine.kt
package com.baseapp.android.view.games.astro

import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.hypot
import kotlin.math.min
import kotlin.math.sin
import kotlin.random.Random

/**
 * Pure Kotlin port of the web AstroBot physics
 * (`web/app/src/View/Games/Astro/AstroGame.tsx` + `entities.ts`) so the exact
 * same rules run on every platform and can be unit-tested on the JVM. No
 * Android or Compose imports here — rendering and input live in
 * `AstroGameView`.
 *
 * Controls arrive as held-input flags ([isTurningLeft] etc.) — the native
 * twin of the web's pressed-keys set — so the view layer only translates
 * touches into booleans and never touches the rules.
 *
 * All coordinates are in field units (820x620, y growing downward); the view
 * scales the field to the device while preserving the aspect ratio. Time is
 * the engine's own [elapsed] clock (the web uses `performance.now()`), so
 * identical steps always produce identical states.
 *
 * Randomness (asteroid spawn position/speed/shape, debris spray) goes through
 * an injected [Random] so tests can make the simulation fully deterministic.
 */
class AstroEngine(private val random: Random = Random.Default) {
    /**
     * Asteroid size tiers: radius (px), base drift speed (px/s, scaled
     * 0.7–1.3x at spawn) and score — the web `radiusByTier` / `speedByTier` /
     * `SCORE_BY_TIER` tables exactly.
     */
    enum class Tier(val radius: Float, val speed: Float, val score: Int) {
        SMALL(15f, 110f, 100),
        MEDIUM(28f, 70f, 50),
        LARGE(46f, 40f, 20);

        /** Tier the two split children spawn at; null rocks vaporize outright. */
        val smaller: Tier?
            get() = when (this) {
                LARGE -> MEDIUM
                MEDIUM -> SMALL
                SMALL -> null
            }
    }

    /** The player ship; `angle` is the heading in radians (-π/2 points up). */
    class Ship(
        var x: Float,
        var y: Float,
        var angle: Float,
        var vx: Float = 0f,
        var vy: Float = 0f,
        /** Engine-clock time (s) the respawn shield drops. */
        var invulnerableUntil: Float = 0f,
        var thrusting: Boolean = false,
    )

    /**
     * One rock. `spin`/`rotation`/`lumps` are the baked-once lumpy silhouette
     * from the web `spawnAsteroidAt`, carried in the engine because they are
     * part of the deterministic RNG stream.
     */
    class Asteroid(
        var x: Float,
        var y: Float,
        var vx: Float,
        var vy: Float,
        val tier: Tier,
        val spin: Float,
        var rotation: Float = 0f,
        /** Radius multipliers (0.75–1.25) for each silhouette vertex. */
        val lumps: List<Float>,
    ) {
        val radius: Float get() = tier.radius
    }

    class Bullet(
        var x: Float,
        var y: Float,
        var vx: Float,
        var vy: Float,
        /** Engine-clock time (s) the bullet was fired. */
        val bornAt: Float,
    )

    /**
     * Explosion debris color family — the web passes literal hex colors into
     * `makeExplosion`; the engine only records which palette entry to use.
     */
    enum class ParticleKind { ASTEROID_DEBRIS, SHIP_DEBRIS }

    class Particle(
        var x: Float,
        var y: Float,
        var vx: Float,
        var vy: Float,
        /** Remaining life in seconds; the renderer uses it as the fade alpha. */
        var life: Float,
        val kind: ParticleKind,
    )

    /**
     * Discrete things that happened during one [step] — the native twin of
     * the web game's HUD/game-over callbacks. Null/empty fields mean
     * "nothing happened".
     */
    data class StepResult(
        val destroyedTiers: List<Tier> = emptyList(),
        /** Lives remaining after a ship-destroying hit, or null if unharmed. */
        val shipDestroyedLivesLeft: Int? = null,
        /** The new sector number when the field was cleared this step. */
        val sectorClearedLevel: Int? = null,
        val isGameOver: Boolean = false,
    )

    // Held inputs, set by the view every time a touch begins/ends. These are
    // the native equivalent of the web's `keys` set (←/→, ↑, space).
    var isTurningLeft = false
    var isTurningRight = false
    var isThrusting = false
    var isFiring = false

    /**
     * Engine clock in seconds, advanced by [step]. Replaces the web's
     * `performance.now()` so the simulation stays pure and step-driven.
     * (Declared before [ship]: `freshShip()` reads it.)
     */
    var elapsed: Float = 0f
        private set

    var ship: Ship = freshShip()
        private set

    private val mutableAsteroids = mutableListOf<Asteroid>()
    val asteroids: List<Asteroid> get() = mutableAsteroids

    private val mutableBullets = mutableListOf<Bullet>()
    val bullets: List<Bullet> get() = mutableBullets

    private val mutableParticles = mutableListOf<Particle>()
    val particles: List<Particle> get() = mutableParticles

    var score: Int = 0
        private set
    var lives: Int = STARTING_LIVES
        private set
    var level: Int = 1
        private set
    var isOver: Boolean = false
        private set

    private var lastShotAt = Float.NEGATIVE_INFINITY

    val isShipInvulnerable: Boolean get() = elapsed < ship.invulnerableUntil

    init {
        startLevel(1)
    }

    /** Full reset; the web equivalent is the resetToken effect. */
    fun newGame() {
        elapsed = 0f
        score = 0
        lives = STARTING_LIVES
        isOver = false
        mutableBullets.clear()
        mutableParticles.clear()
        lastShotAt = Float.NEGATIVE_INFINITY
        ship = freshShip()
        startLevel(1)
    }

    /**
     * Advance the simulation by [dtSeconds] (clamped to 50ms like the web
     * loop, so a background pause never produces a huge tunnel-through
     * step). Mirrors the web `step` frame-for-frame: turn, thrust, drag,
     * move+wrap ship, fire, move bullets/asteroids/particles,
     * bullet↔asteroid hits (with splitting), asteroid↔ship collision, and
     * the next-sector check.
     */
    fun step(dtSeconds: Float): StepResult {
        if (isOver) {
            return StepResult()
        }
        val dt = min(dtSeconds, MAX_STEP_SECONDS)
        elapsed += dt
        val now = elapsed

        // Ship: rotate, thrust, drag, integrate, wrap (margin = ship radius).
        if (isTurningLeft) {
            ship.angle -= TURN_SPEED * dt
        }
        if (isTurningRight) {
            ship.angle += TURN_SPEED * dt
        }
        if (isThrusting) {
            ship.vx += cos(ship.angle) * THRUST * dt
            ship.vy += sin(ship.angle) * THRUST * dt
        }
        ship.vx *= 1 - DRAG * dt
        ship.vy *= 1 - DRAG * dt
        ship.x += ship.vx * dt
        ship.y += ship.vy * dt
        ship.thrusting = isThrusting
        ship.x = wrapX(ship.x, SHIP_RADIUS)
        ship.y = wrapY(ship.y, SHIP_RADIUS)

        // Fire from the nose, inheriting ship velocity (web makeBullet).
        if (isFiring && now - lastShotAt > FIRE_COOLDOWN_SECONDS) {
            lastShotAt = now
            mutableBullets.add(
                Bullet(
                    x = ship.x + cos(ship.angle) * SHIP_RADIUS,
                    y = ship.y + sin(ship.angle) * SHIP_RADIUS,
                    vx = ship.vx + cos(ship.angle) * BULLET_SPEED,
                    vy = ship.vy + sin(ship.angle) * BULLET_SPEED,
                    bornAt = now,
                ),
            )
        }

        // Bullets: expire, move, wrap (no margin, like web).
        mutableBullets.removeAll { now - it.bornAt >= BULLET_LIFE_SECONDS }
        for (bullet in mutableBullets) {
            bullet.x += bullet.vx * dt
            bullet.y += bullet.vy * dt
            bullet.x = wrapX(bullet.x, 0f)
            bullet.y = wrapY(bullet.y, 0f)
        }

        // Asteroids: drift, spin, wrap (margin = own radius).
        for (asteroid in mutableAsteroids) {
            asteroid.x += asteroid.vx * dt
            asteroid.y += asteroid.vy * dt
            asteroid.rotation += asteroid.spin * dt
            asteroid.x = wrapX(asteroid.x, asteroid.radius)
            asteroid.y = wrapY(asteroid.y, asteroid.radius)
        }

        // Particles: age out, survivors drift (web filters then moves).
        for (particle in mutableParticles) {
            particle.life -= dt
        }
        mutableParticles.removeAll { it.life <= 0f }
        for (particle in mutableParticles) {
            particle.x += particle.vx * dt
            particle.y += particle.vy * dt
        }

        // Bullet -> asteroid hits: the first overlapping bullet is consumed,
        // the rock scores and splits until SMALL (two children per split).
        val destroyedTiers = mutableListOf<Tier>()
        val survivors = mutableListOf<Asteroid>()
        for (asteroid in mutableAsteroids) {
            val hit = mutableBullets.firstOrNull {
                hypot(it.x - asteroid.x, it.y - asteroid.y) < asteroid.radius
            }
            if (hit == null) {
                survivors.add(asteroid)
                continue
            }
            mutableBullets.remove(hit)
            score += asteroid.tier.score
            spawnExplosion(asteroid.x, asteroid.y, ParticleKind.ASTEROID_DEBRIS)
            destroyedTiers.add(asteroid.tier)
            asteroid.tier.smaller?.let { childTier ->
                survivors.add(spawnAsteroidAt(asteroid.x, asteroid.y, childTier))
                survivors.add(spawnAsteroidAt(asteroid.x, asteroid.y, childTier))
            }
        }
        mutableAsteroids.clear()
        mutableAsteroids.addAll(survivors)

        // Asteroid -> ship collision (web pads the hit circle by -4 px to be
        // forgiving). Skipped entirely while the respawn shield is up.
        var shipDestroyedLivesLeft: Int? = null
        if (!isShipInvulnerable) {
            val collided = mutableAsteroids.any {
                hypot(it.x - ship.x, it.y - ship.y) < it.radius + SHIP_RADIUS - 4f
            }
            if (collided) {
                lives -= 1
                spawnExplosion(ship.x, ship.y, ParticleKind.SHIP_DEBRIS)
                if (lives <= 0) {
                    isOver = true
                    return StepResult(
                        destroyedTiers = destroyedTiers,
                        shipDestroyedLivesLeft = 0,
                        isGameOver = true,
                    )
                }
                shipDestroyedLivesLeft = lives
                ship = freshShip()
            }
        }

        // Field clear: next sector spawns immediately plus the +250 bonus.
        var sectorClearedLevel: Int? = null
        if (mutableAsteroids.isEmpty()) {
            startLevel(level + 1)
            score += SECTOR_BONUS
            sectorClearedLevel = level
        }

        return StepResult(
            destroyedTiers = destroyedTiers,
            shipDestroyedLivesLeft = shipDestroyedLivesLeft,
            sectorClearedLevel = sectorClearedLevel,
        )
    }

    /** Fresh centered ship with the 2s shield (web `makeShip`). */
    private fun freshShip(): Ship = Ship(
        x = FIELD_WIDTH / 2f,
        y = FIELD_HEIGHT / 2f,
        angle = (-PI / 2).toFloat(),
        invulnerableUntil = elapsed + RESPAWN_INVULNERABILITY_SECONDS,
    )

    /**
     * Populate a sector: `2 + level` large rocks kept away from the ship
     * (web `startLevel` + `makeAsteroid`). The web rejection-samples with an
     * unbounded do/while; we cap the attempts so a pathological RNG can
     * never hang the engine.
     */
    private fun startLevel(newLevel: Int) {
        level = newLevel
        mutableAsteroids.clear()
        repeat(2 + newLevel) {
            var x = random.nextFloat() * FIELD_WIDTH
            var y = random.nextFloat() * FIELD_HEIGHT
            var attempts = 0
            while (hypot(x - ship.x, y - ship.y) < SAFE_SPAWN_DISTANCE && attempts < 100) {
                x = random.nextFloat() * FIELD_WIDTH
                y = random.nextFloat() * FIELD_HEIGHT
                attempts += 1
            }
            mutableAsteroids.add(spawnAsteroidAt(x, y, Tier.LARGE))
        }
    }

    /**
     * Web `spawnAsteroidAt`: random heading, 0.7–1.3x tier speed, random
     * spin and a baked lumpy silhouette of 9–12 vertices. RNG call order
     * matches the web factory so seeded runs stay comparable.
     */
    private fun spawnAsteroidAt(x: Float, y: Float, tier: Tier): Asteroid {
        val angle = random.nextFloat() * PI.toFloat() * 2f
        val speed = tier.speed * (0.7f + random.nextFloat() * 0.6f)
        val pointCount = 9 + (random.nextFloat() * 4f).toInt()
        return Asteroid(
            x = x,
            y = y,
            vx = cos(angle) * speed,
            vy = sin(angle) * speed,
            tier = tier,
            spin = (random.nextFloat() - 0.5f) * 1.6f,
            lumps = List(pointCount) { 0.75f + random.nextFloat() * 0.5f },
        )
    }

    /** Web `makeExplosion`: 14 debris sparks with random spread and lifetime. */
    private fun spawnExplosion(x: Float, y: Float, kind: ParticleKind) {
        repeat(14) {
            val angle = random.nextFloat() * PI.toFloat() * 2f
            val speed = 40f + random.nextFloat() * 140f
            mutableParticles.add(
                Particle(
                    x = x,
                    y = y,
                    vx = cos(angle) * speed,
                    vy = sin(angle) * speed,
                    life = 0.5f + random.nextFloat() * 0.4f,
                    kind = kind,
                ),
            )
        }
    }

    /** Toroidal wrap for the x axis with a per-entity margin (web `wrap`). */
    private fun wrapX(x: Float, margin: Float): Float = when {
        x < -margin -> FIELD_WIDTH + margin
        x > FIELD_WIDTH + margin -> -margin
        else -> x
    }

    /** Toroidal wrap for the y axis with a per-entity margin (web `wrap`). */
    private fun wrapY(y: Float, margin: Float): Float = when {
        y < -margin -> FIELD_HEIGHT + margin
        y > FIELD_HEIGHT + margin -> -margin
        else -> y
    }

    // Test hooks (internal, like PongEngine.placeBall): build exact scenarios.

    /**
     * Test hook: clear the rock field. (The next [step] will treat an empty
     * field as a cleared sector.)
     */
    internal fun removeAllAsteroids() {
        mutableAsteroids.clear()
    }

    /** Test hook: spawn a rock at an exact spot with an exact velocity. */
    internal fun addAsteroid(x: Float, y: Float, tier: Tier, vx: Float = 0f, vy: Float = 0f) {
        val asteroid = spawnAsteroidAt(x, y, tier)
        asteroid.vx = vx
        asteroid.vy = vy
        mutableAsteroids.add(asteroid)
    }

    /** Test hook: place a live bullet directly. */
    internal fun addBullet(x: Float, y: Float, vx: Float = 0f, vy: Float = 0f) {
        mutableBullets.add(Bullet(x, y, vx, vy, bornAt = elapsed))
    }

    /** Test hook: teleport the ship. */
    internal fun placeShip(x: Float, y: Float, vx: Float = 0f, vy: Float = 0f) {
        ship.x = x
        ship.y = y
        ship.vx = vx
        ship.vy = vy
    }

    /** Test hook: drop the respawn shield immediately. */
    internal fun endInvulnerability() {
        ship.invulnerableUntil = -1f
    }

    companion object {
        // Field geometry and tuning — must stay byte-for-byte in sync with
        // the web constants in AstroGame.tsx / entities.ts.
        const val FIELD_WIDTH = 820f
        const val FIELD_HEIGHT = 620f
        const val TURN_SPEED = 3.6f
        const val THRUST = 300f
        const val DRAG = 0.4f

        /** Web FIRE_COOLDOWN_MS = 180. */
        const val FIRE_COOLDOWN_SECONDS = 0.18f

        /** Web BULLET_LIFE_MS = 1100. */
        const val BULLET_LIFE_SECONDS = 1.1f
        const val BULLET_SPEED = 460f
        const val STARTING_LIVES = 3
        const val SHIP_RADIUS = 14f

        /** Fresh ships are safe for 2s (web `makeShip`: now + 2000). */
        const val RESPAWN_INVULNERABILITY_SECONDS = 2f

        /** Bonus for clearing a sector. */
        const val SECTOR_BONUS = 250

        /** New rocks never spawn within this distance of the ship (web: 160). */
        const val SAFE_SPAWN_DISTANCE = 160f
        const val MAX_STEP_SECONDS = 0.05f
    }
}

File: ios/App/View/Games/Astro/AstroEngine.swift
import Foundation

/// Asteroid size tiers. `radius` (px), base `speed` (px/s) and `score` mirror
/// the web `radiusByTier` / `speedByTier` / `SCORE_BY_TIER` tables exactly so
/// both platforms play and score identically.
enum AstroTier: Int, CaseIterable, Equatable {
  case small = 1
  case medium = 2
  case large = 3

  var radius: Double {
    switch self {
    case .small: return 15
    case .medium: return 28
    case .large: return 46
    }
  }

  /// Base drift speed; the actual speed is scaled by 0.7–1.3 at spawn,
  /// exactly like the web `spawnAsteroidAt`.
  var speed: Double {
    switch self {
    case .small: return 110
    case .medium: return 70
    case .large: return 40
    }
  }

  var score: Int {
    switch self {
    case .small: return 100
    case .medium: return 50
    case .large: return 20
    }
  }

  /// Tier the two split children spawn at; nil for the smallest rocks, which
  /// vaporize outright.
  var smaller: AstroTier? { AstroTier(rawValue: rawValue - 1) }
}

/// The player ship. `invulnerableUntil` is in engine seconds (the web stores
/// `performance.now() + 2000`; the engine keeps its own clock instead so the
/// simulation stays pure and step-driven).
struct AstroShip: Equatable {
  var x: Double
  var y: Double
  /// Heading in radians; -π/2 points "up" like the web ship.
  var angle: Double
  var vx: Double
  var vy: Double
  var invulnerableUntil: Double
  var thrusting: Bool
}

/// One rock. `spin`/`rotation`/`lumps` are the baked-once visual identity
/// from the web `spawnAsteroidAt` (lumpy silhouette), carried in the engine
/// because they are part of the deterministic RNG stream.
struct AstroAsteroid: Equatable {
  var x: Double
  var y: Double
  var vx: Double
  var vy: Double
  let tier: AstroTier
  let radius: Double
  let spin: Double
  var rotation: Double
  /// Radius multipliers (0.75–1.25) for each silhouette vertex.
  let lumps: [Double]
}

struct AstroBullet: Equatable {
  var x: Double
  var y: Double
  var vx: Double
  var vy: Double
  /// Engine-clock time the bullet was fired (web: `performance.now()`).
  var bornAt: Double
}

/// Explosion debris color family — the web passes literal hex colors into
/// `makeExplosion`; the engine only records which palette entry to use.
enum AstroParticleKind: Equatable {
  /// Web "#ff9d5c" — asteroid destruction.
  case asteroidDebris
  /// Web "#57c8ff" — ship destruction.
  case shipDebris
}

struct AstroParticle: Equatable {
  var x: Double
  var y: Double
  var vx: Double
  var vy: Double
  /// Remaining life in seconds; the renderer uses it as the fade alpha.
  var life: Double
  let kind: AstroParticleKind
}

/// Discrete things that happened during one `step(dt:)` — the native twin of
/// the web game's HUD/game-over callbacks. The renderer can use these for
/// sounds or haptics; tests use them to assert on game flow.
enum AstroEvent: Equatable {
  case asteroidDestroyed(tier: AstroTier)
  case shipDestroyed(livesLeft: Int)
  case sectorCleared(newLevel: Int)
  case gameOver(score: Int, level: Int)
}

/// Pure port of the web AstroBot simulation
/// (`web/app/src/View/Games/Astro/AstroGame.tsx` + `entities.ts`). No SwiftUI
/// here: the engine is a plain state machine driven by `step(dt:)` so it can
/// be unit-tested headlessly and rendered by any frontend.
///
/// Controls arrive as held-input flags (`isTurningLeft` etc.) — the native
/// twin of the web's pressed-keys set — so the view layer only translates
/// touches into booleans and never touches the rules.
///
/// Randomness (asteroid spawn position/speed/shape, debris spray) goes
/// through an injected closure so tests can make the simulation fully
/// deterministic.
final class AstroEngine {
  // Field geometry and tuning — must stay byte-for-byte in sync with the web
  // constants in AstroGame.tsx / entities.ts.
  static let fieldWidth: Double = 820
  static let fieldHeight: Double = 620
  static let turnSpeed: Double = 3.6
  static let thrust: Double = 300
  static let drag: Double = 0.4
  /// Web FIRE_COOLDOWN_MS = 180.
  static let fireCooldown: Double = 0.18
  /// Web BULLET_LIFE_MS = 1100.
  static let bulletLife: Double = 1.1
  static let bulletSpeed: Double = 460
  static let startingLives = 3
  static let shipRadius: Double = 14
  /// Fresh ships are safe for 2s (web `makeShip`: now + 2000).
  static let respawnInvulnerability: Double = 2.0
  /// Bonus for clearing a sector.
  static let sectorBonus = 250
  /// New rocks never spawn within this distance of the ship (web: 160).
  static let safeSpawnDistance: Double = 160

  // Held inputs, set by the view every time a touch begins/ends. These are
  // the native equivalent of the web's `keys` set (←/→, ↑, space).
  var isTurningLeft = false
  var isTurningRight = false
  var isThrusting = false
  var isFiring = false

  private(set) var ship: AstroShip
  private(set) var asteroids: [AstroAsteroid] = []
  private(set) var bullets: [AstroBullet] = []
  private(set) var particles: [AstroParticle] = []
  private(set) var score = 0
  private(set) var lives = AstroEngine.startingLives
  private(set) var level = 1
  private(set) var isOver = false
  /// Engine clock in seconds, advanced by `step(dt:)`. Replaces the web's
  /// `performance.now()` so identical steps always produce identical states.
  private(set) var elapsed: Double = 0

  private var lastShotAt = -Double.greatestFiniteMagnitude
  private let random: () -> Double

  var isShipInvulnerable: Bool { elapsed < ship.invulnerableUntil }

  init(random: @escaping () -> Double = { Double.random(in: 0..<1) }) {
    self.random = random
    self.ship = AstroShip(
      x: Self.fieldWidth / 2, y: Self.fieldHeight / 2, angle: -Double.pi / 2,
      vx: 0, vy: 0, invulnerableUntil: Self.respawnInvulnerability, thrusting: false
    )
    startLevel(1)
  }

  /// Full reset (the web "New Game" / resetToken path).
  func newGame() {
    elapsed = 0
    score = 0
    lives = Self.startingLives
    isOver = false
    bullets = []
    particles = []
    lastShotAt = -Double.greatestFiniteMagnitude
    ship = makeShip()
    startLevel(1)
  }

  /// Advance the simulation. Mirrors the web `step` frame-for-frame: turn,
  /// thrust, drag, move+wrap ship, fire, move bullets/asteroids/particles,
  /// bullet↔asteroid hits (with splitting), asteroid↔ship collision, and the
  /// next-sector check. Returns the discrete events that occurred.
  func step(dt: Double) -> [AstroEvent] {
    guard !isOver else { return [] }
    var events: [AstroEvent] = []
    elapsed += dt
    let now = elapsed

    // Ship: rotate, thrust, drag, integrate, wrap (web margin = ship radius).
    if isTurningLeft { ship.angle -= Self.turnSpeed * dt }
    if isTurningRight { ship.angle += Self.turnSpeed * dt }
    if isThrusting {
      ship.vx += cos(ship.angle) * Self.thrust * dt
      ship.vy += sin(ship.angle) * Self.thrust * dt
    }
    ship.vx *= 1 - Self.drag * dt
    ship.vy *= 1 - Self.drag * dt
    ship.x += ship.vx * dt
    ship.y += ship.vy * dt
    ship.thrusting = isThrusting
    (ship.x, ship.y) = wrapped(x: ship.x, y: ship.y, margin: Self.shipRadius)

    // Fire from the nose, inheriting ship velocity (web makeBullet).
    if isFiring, now - lastShotAt > Self.fireCooldown {
      lastShotAt = now
      bullets.append(AstroBullet(
        x: ship.x + cos(ship.angle) * Self.shipRadius,
        y: ship.y + sin(ship.angle) * Self.shipRadius,
        vx: ship.vx + cos(ship.angle) * Self.bulletSpeed,
        vy: ship.vy + sin(ship.angle) * Self.bulletSpeed,
        bornAt: now
      ))
    }

    // Bullets: expire, move, wrap (no margin, like web).
    bullets.removeAll { now - $0.bornAt >= Self.bulletLife }
    for index in bullets.indices {
      var bullet = bullets[index]
      bullet.x += bullet.vx * dt
      bullet.y += bullet.vy * dt
      (bullet.x, bullet.y) = wrapped(x: bullet.x, y: bullet.y, margin: 0)
      bullets[index] = bullet
    }

    // Asteroids: drift, spin, wrap (margin = own radius).
    for index in asteroids.indices {
      var asteroid = asteroids[index]
      asteroid.x += asteroid.vx * dt
      asteroid.y += asteroid.vy * dt
      asteroid.rotation += asteroid.spin * dt
      (asteroid.x, asteroid.y) = wrapped(x: asteroid.x, y: asteroid.y, margin: asteroid.radius)
      asteroids[index] = asteroid
    }

    // Particles: age out, survivors drift (web filters then moves).
    for index in particles.indices {
      particles[index].life -= dt
    }
    particles.removeAll { $0.life <= 0 }
    for index in particles.indices {
      particles[index].x += particles[index].vx * dt
      particles[index].y += particles[index].vy * dt
    }

    // Bullet -> asteroid hits: first overlapping bullet is consumed, the rock
    // scores and splits until tier 1 (two children per split, like web).
    var survivors: [AstroAsteroid] = []
    for asteroid in asteroids {
      let hitIndex = bullets.firstIndex {
        hypot($0.x - asteroid.x, $0.y - asteroid.y) < asteroid.radius
      }
      guard let hitIndex else {
        survivors.append(asteroid)
        continue
      }
      bullets.remove(at: hitIndex)
      score += asteroid.tier.score
      particles.append(contentsOf: makeExplosion(x: asteroid.x, y: asteroid.y, kind: .asteroidDebris))
      events.append(.asteroidDestroyed(tier: asteroid.tier))
      if let smaller = asteroid.tier.smaller {
        survivors.append(spawnAsteroid(x: asteroid.x, y: asteroid.y, tier: smaller))
        survivors.append(spawnAsteroid(x: asteroid.x, y: asteroid.y, tier: smaller))
      }
    }
    asteroids = survivors

    // Asteroid -> ship collision (web pads the hit circle by -4 px to be
    // forgiving). Skipped entirely while the respawn shield is up.
    if !isShipInvulnerable {
      let collided = asteroids.contains {
        hypot($0.x - ship.x, $0.y - ship.y) < $0.radius + Self.shipRadius - 4
      }
      if collided {
        lives -= 1
        particles.append(contentsOf: makeExplosion(x: ship.x, y: ship.y, kind: .shipDebris))
        if lives <= 0 {
          isOver = true
          events.append(.gameOver(score: score, level: level))
        } else {
          events.append(.shipDestroyed(livesLeft: lives))
          ship = makeShip()
        }
      }
    }

    // Field clear: next sector spawns immediately plus the +250 bonus.
    if asteroids.isEmpty {
      startLevel(level + 1)
      score += Self.sectorBonus
      events.append(.sectorCleared(newLevel: level))
    }

    return events
  }

  // MARK: - Test hooks

  /// Test hook: clear the rock field to build exact collision scenarios.
  /// (The next `step` will treat an empty field as a cleared sector.)
  func removeAllAsteroids() {
    asteroids = []
  }

  /// Test hook: spawn a rock at an exact spot with an exact velocity.
  func addAsteroid(x: Double, y: Double, tier: AstroTier, vx: Double = 0, vy: Double = 0) {
    var asteroid = spawnAsteroid(x: x, y: y, tier: tier)
    asteroid.vx = vx
    asteroid.vy = vy
    asteroids.append(asteroid)
  }

  /// Test hook: place a live bullet directly.
  func addBullet(x: Double, y: Double, vx: Double = 0, vy: Double = 0) {
    bullets.append(AstroBullet(x: x, y: y, vx: vx, vy: vy, bornAt: elapsed))
  }

  /// Test hook: teleport the ship.
  func setShip(x: Double, y: Double, vx: Double = 0, vy: Double = 0, angle: Double = -Double.pi / 2) {
    ship.x = x
    ship.y = y
    ship.vx = vx
    ship.vy = vy
    ship.angle = angle
  }

  /// Test hook: drop the respawn shield immediately.
  func endInvulnerability() {
    ship.invulnerableUntil = -1
  }

  // MARK: - Spawning

  /// Fresh centered ship with the 2s shield (web `makeShip`).
  private func makeShip() -> AstroShip {
    AstroShip(
      x: Self.fieldWidth / 2, y: Self.fieldHeight / 2, angle: -Double.pi / 2,
      vx: 0, vy: 0, invulnerableUntil: elapsed + Self.respawnInvulnerability,
      thrusting: false
    )
  }

  /// Populate a sector: `2 + level` large rocks kept away from the ship
  /// (web `startLevel` + `makeAsteroid`).
  private func startLevel(_ newLevel: Int) {
    level = newLevel
    asteroids = (0..<(2 + newLevel)).map { _ in
      // The web rejection-samples with an unbounded do/while; we cap the
      // attempts so a pathological RNG can never hang the engine.
      var x = random() * Self.fieldWidth
      var y = random() * Self.fieldHeight
      for _ in 0..<100 where hypot(x - ship.x, y - ship.y) < Self.safeSpawnDistance {
        x = random() * Self.fieldWidth
        y = random() * Self.fieldHeight
      }
      return spawnAsteroid(x: x, y: y, tier: .large)
    }
  }

  /// Web `spawnAsteroidAt`: random heading, 0.7–1.3× tier speed, random spin
  /// and a baked lumpy silhouette of 9–12 vertices. RNG call order matches
  /// the web factory so seeded runs stay comparable.
  private func spawnAsteroid(x: Double, y: Double, tier: AstroTier) -> AstroAsteroid {
    let angle = random() * Double.pi * 2
    let speed = tier.speed * (0.7 + random() * 0.6)
    let pointCount = 9 + Int(random() * 4)
    return AstroAsteroid(
      x: x,
      y: y,
      vx: cos(angle) * speed,
      vy: sin(angle) * speed,
      tier: tier,
      radius: tier.radius,
      spin: (random() - 0.5) * 1.6,
      rotation: 0,
      lumps: (0..<pointCount).map { _ in 0.75 + random() * 0.5 }
    )
  }

  /// Web `makeExplosion`: 14 debris sparks with random spread and lifetime.
  private func makeExplosion(x: Double, y: Double, kind: AstroParticleKind) -> [AstroParticle] {
    (0..<14).map { _ in
      let angle = random() * Double.pi * 2
      let speed = 40 + random() * 140
      return AstroParticle(
        x: x,
        y: y,
        vx: cos(angle) * speed,
        vy: sin(angle) * speed,
        life: 0.5 + random() * 0.4,
        kind: kind
      )
    }
  }

  /// Toroidal wrap with a per-entity margin (web `wrap`).
  private func wrapped(x: Double, y: Double, margin: Double) -> (Double, Double) {
    var x = x
    var y = y
    if x < -margin { x = Self.fieldWidth + margin }
    if x > Self.fieldWidth + margin { x = -margin }
    if y < -margin { y = Self.fieldHeight + margin }
    if y > Self.fieldHeight + margin { y = -margin }
    return (x, y)
  }
}
